mkdocs/config/config_options.py

```diff
@@ -0,0 +1,333 @@
+                raise ValidationError("Required configuration not provided.")
+
+        return self._validate(value)
+
+    def _validate(self, value):
+        """
+        Perform validation for a value.
+
+        The _validate method should be implemented by subclasses.
+        """
+        return value
+
+    def post_process(self, config, key_name):
+        """
+        After all options have passed validation, perform a post process to
+        do any additional changes dependant on other config values.
+
+        The post process method should be implemented by subclasses.
+        """
+
+
+class Type(BaseConfigOption):
+    """
+    Type Config Option
+
+    Validate the type of a config option against a given Python type.
+    """
+
+    def __init__(self, type_, length=None, **kwargs):
+        super(Type, self).__init__(**kwargs)
+        self._type = type_
+        self.length = length
+
+    def _validate(self, value):
+
+        if not isinstance(value, self._type):
+            msg = ("Expected type: {0} but recieved: {1}"
+                   .format(self._type, type(value)))
+        elif self.length is not None and len(value) != self.length:
+            msg = ("Expected type: {0} with lenght {2} but recieved: {1} with "
+                   "length {3}").format(self._type, value, self.length,
+                                        len(value))
+        else:
+            return value
+
+        raise ValidationError(msg)
+
+
+class URL(BaseConfigOption):
+    """
+    URL Config Option
+
+    Validate a URL by requiring a scheme is present.
+    """
+
+    def _validate(self, value):
+
+        parsed_url = six.moves.urllib.parse.urlparse(value)
+
+        if parsed_url.scheme:
+            return value
+
+        raise ValidationError("Invalid URL.")
+
+
+class RepoURL(URL):
+    """
+    Repo URL Config Option
+
+    A small extension to the URL config that sets the repo_name, based on the
+    url if it hasn't already been provided.
+    """
+
+    def post_process(self, config, key_name):
+
+        if config['repo_url'] is not None and config.get('repo_name') is None:
+            repo_host = six.moves.urllib.parse.urlparse(
+                config['repo_url']).netloc.lower()
+            if repo_host == 'github.com':
+                config['repo_name'] = 'GitHub'
+            elif repo_host == 'bitbucket.org':
+                config['repo_name'] = 'Bitbucket'
+            else:
+                config['repo_name'] = repo_host.split('.')[0].title()
+
+
+class Dir(BaseConfigOption):
+    """
+    Dir Config Option
+
+    Validate a path to a directory, optionally verifying that it exists.
+    """
+
+    def __init__(self, exists=False, **kwargs):
+        super(Dir, self).__init__(**kwargs)
+        self.exists = exists
+
+    def _validate(self, value):
+
+        if self.exists and not os.path.isdir(value):
+            raise ValidationError("The path {0} doesn't exist".format(value))
```

Comment:
Makes sense, I'll do both. Some of this didn't happen as I was trying to draw a line on this piece of work :) There is no validation for this at the moment, I expect to improve them all over time somewhat, but I should add a better baseline here.